jupyterhub/tests/test_scopes.py

```diff
@@ -189,11 +228,105 @@ async def test_expand_groups(app, user_name, in_group, status_code):
         app.db.add(group)
     if in_group and user not in group.users:
         group.users.append(user)
+    kind = 'users'
+    roles.update_roles(app.db, user, kind, roles=['test'])
+    roles.remove_obj(app.db, user_name, kind, 'user')
     app.db.commit()
-    app.tornado_settings['mock_scopes'] = [
-        'read:users!user=martha',
-        'read:users!group=bluth',
-        'read:groups',
-    ]
-    r = await api_request(app, 'users', user_name)
+    r = await api_request(
+        app, 'users', user_name, headers=auth_header(app.db, user_name)
+    )
     assert r.status_code == status_code
+
+
+async def test_user_filter(app):
+    user_name = 'rita'
+    test_role = {
+        'name': 'test',
+        'description': '',
+        'users': [user_name],
+        'scopes': [
+            'read:users!user=lindsay',
+            'read:users!user=gob',
+            'read:users!user=oscar',
+        ],
+    }
+    roles.add_role(app.db, test_role)
+    name_in_scope = {'lindsay', 'oscar', 'gob'}
+    outside_scope = {'maeby', 'marta'}
+    group_name = 'bluth'
+    group = orm.Group.find(app.db, name=group_name)
+    if not group:
+        group = orm.Group(name=group_name)
+        app.db.add(group)
+    for name in name_in_scope | outside_scope:
+        user = add_user(app.db, name=name)
+        if name not in group.users:
+            group.users.append(user)
+    kind = 'users'
+    user = add_user(app.db, name=user_name)
+    roles.update_roles(app.db, user, kind, roles=['test'])
+    roles.remove_obj(app.db, user_name, kind, 'user')
+    app.db.commit()
+    r = await api_request(app, 'users', headers=auth_header(app.db, user_name))
+    assert r.status_code == 200
+    result_names = {user['name'] for user in r.json()}
+    assert result_names == name_in_scope
+
+
+async def test_user_filter_with_group(app):  # todo: Move role setup to setup method
+    user_name = 'sally'  # Fixme: fails randomly? scopes not always loaded?
+    test_role = {
+        'name': 'test',
+        'description': '',
+        'users': [user_name],
+        'scopes': ['read:users!group=sitwell'],
+    }
+    roles.add_role(app.db, test_role)
+    user = add_user(app.db, name=user_name)
+    name_set = {'sally', 'stan'}
+    group_name = 'sitwell'
+    group = orm.Group.find(app.db, name=group_name)
+    if not group:
+        group = orm.Group(name=group_name)
+        app.db.add(group)
+    for name in name_set:
+        user = add_user(app.db, name=name)
```

Comment:
`user` variable is redefined here and will sometimes be `stan` and sometimes `sally` after this loop, meaning the `update_roles` call below is sometimes called for sally and sometimes for stan. Python's forced iteration randomization of unordered collections means that iterating through `name_set` is *explicitly* nondeterministic. The result should be expected failures ~50% of the time. Renaming this variable to avoid colliding with `user` should solve the issue.